### PR TITLE
[Bug] Reject ES-incompatible field comparisons under elasticsearch_syntax

### PR DESCRIPTION
**Pull request by Mikaayenson**
## Summary
- Under `elasticsearch_syntax` / `reject_field_comparisons`, reject comparisons where **neither side folds to a constant** (field-to-field), including through functions and arithmetic (e.g. `a == b`, `length(a) == length(b) - 1`).
- Comparing a field expression to a **constant** remains valid (e.g. `length(a) - 1 == 5`), matching Elasticsearch stack behavior.
- Not applied under `elastic_endpoint_syntax` (Endpoint allows field-to-field comparisons). Sequence aliases (`as p0`) remain allowed when aliases are enabled.
- Motivated by detection-rules shipping an invalid query that Python EQL accepted but ES rejected (`Comparisons against fields are not currently supported`) ([detection-rules#6611](https://github.com/elastic/detection-rules/pull/6611)).

## Test plan
- [x] `flake8 eql tests`
- [x] Full `pytest tests/` locally (616 passed)
- [x] Edge-case matrix for ES allow/forbid + endpoint allow (25/25)
- [x] Scanned detection-rules `8.19`/`9.3`/`9.4`/`9.5`: only NTLM field-vs-field fails (expected); fixed branch parses clean
- [x] Scanned endpoint-rules `main`: 1936/1936 OK

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,6 +1,16 @@
 # Event Query Language - Changelog
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 
+# Version 1.0.1
+
+ _Unreleased_
```
Comment (Mikaayenson) [inline CHANGELOG.md]: Will update right before merging

File: eql/parser.py [hunk 2]
```diff
@@ -704,6 +709,71 @@ def string_predicate(self, node):
...
+            return
+
+        left_field, left_math, left_alias = self._classify_comparison_side(left.node)
+        right_field, right_math, right_alias = self._classify_comparison_side(right.node)
+
+        # Sequence aliases (``as p0``) may legally compare across events.
+        if left_alias or right_alias:
+            return
+
+        # Illegal: field vs field, or any side with field arithmetic in the comparison.
```
Comment (eric-forte-elastic) [inline eql/parser.py]: For this, do we know that this is illegal with any side (in effect both), or should it be only when neither side folds to a constant (as long as one does its ok)?

E.g.

Image

Current:

Image

I think expected based on Elastic stack behavior: 

Image

Comment (eric-forte-elastic) [reply eql/parser.py]: Possible patch 

[bugfix.patch](https://github.com/user-attachments/files/30797646/bugfix.patch)

Comment (Mikaayenson) [reply eql/parser.py]: ++ accepted your change

File: eql/parser.py [hunk 3]
```diff
@@ -704,6 +709,69 @@ def string_predicate(self, node):
...
+        # Illegal only when neither side folds to a constant.
+        if not (left_field and right_field):
+            return
+
+        raise self._error(
+            node,
+            "Comparisons against fields are not (currently) supported; offender [{}] in [{}]".format(
+                right.node, op
+            ),
+            cls=EqlSemanticError,
```
Comment (eric-forte-elastic) [inline eql/parser.py]:
```suggestion
            "Comparisons against fields are not (currently) supported; offender [{offender}] in [{op}]",
            cls=EqlSemanticError,
            offender=right.node, op=op,
```

I think this will double format the string as it currently sits. Instead, perhaps pass offender and op as kwargs to _error instead of pre-formatting the message string.

Previously `concat(b, "{")` would be `ValueError`. Now, it would be `EqlSemanticError`.

File: tests/test_parser.py [hunk 4]
```diff
@@ -658,6 +658,51 @@ def test_elasticsearch_flag(self):
...
+
+        # Elastic Endpoint EQL allows field-to-field comparisons (unlike ES/SIEM EQL)
+        with elastic_endpoint_syntax, ignore_missing_functions:
+            parse_query('process where parent_name == process_name')
+            parse_query('process where Target.process.executable != process.executable')
+            parse_query('process where length(host_name) == length(user_name) - 1')
+            parse_query(
+                'api where process.Ext.api.metadata.client_process_id == process.pid'
+            )
+
```
Comment (eric-forte-elastic) [inline tests/test_parser.py:730]: Test cases for alias-to-event comparisons. They are valid in ES EQL.

```suggestion
        # Under elasticsearch_syntax + allow_alias, sequence event aliases (``as p0``) exempt
        # comparisons that reference them, because alias-to-event comparisons are valid in ES EQL.
        es_alias_syntax = ParserConfig(elasticsearch_syntax=True, reject_field_comparisons=True,
                                       allow_alias=True)
        with es_alias_syntax, ignore_missing_functions:
            # Alias-referencing comparisons are allowed
            parse_query(
                'sequence [process where true] as p0 [process where p0.process_name == process_name]'
            )
            parse_query(
                'sequence [process where true] as p0 '
                '[process where length(p0.process_name) == length(process_name) - 1]'
            )
            # Non-alias field-vs-field is still rejected even with aliases present
            self.assertRaises(
                EqlSemanticError, parse_query,
                'sequence [process where true] as p0 [process where parent_name == process_name]'
            )

```

**Review by eric-forte-elastic — APPROVED** [review]
Two suggestions, otherwise I think this looks good 👍

File: eql/parser.py (hunk 3 above)
Comment (Mikaayenson) [reply eql/parser.py]: accepted

File: tests/test_parser.py (hunk 4 above)
Comment (Mikaayenson) [reply tests/test_parser.py:730]: accepted

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,6 +1,18 @@
...
```
Comment (Mikaayenson) [inline CHANGELOG.md]:
```suggestion
 _Released 2026-08-07_
```
